### PR TITLE
Use uv workspaces instead of pip

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

I don't know what I'm doing and this doesn't work.

* The lock file looks okay?
* I can't run any scripts because the dependencies don't seem to get installed
* I'm probably doing something very wrong

## Test Plan

See contributing.md. I didn't get passed `uv sync`

### CODE REVIEW

File: docs/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  "mdformat==0.7.17",
+  "mdformat-mkdocs==2.0.4",
+  "mdformat-admon==2.0.2"
+]
+
+[tool.uv]
+dev-dependencies = [
+  "mkdocs-material==9.1.18",
+  "mkdocs-redirects==1.2.1",
+]
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply docs/pyproject.toml:21]: lol, slightly emberassing :D Thanks

File: docs/requirements-insiders.txt [hunk 2]
```diff
@@ -1,8 +1 @@
-PyYAML==6.0.1
-black==23.10.0
-mkdocs==1.5.0
 mkdocs-material @ git+ssh://git@github.com/astral-sh/mkdocs-material-insiders.git@38c0b8187325c3bab386b666daf3518ac036f2f4
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply docs/requirements-insiders.txt:1]: We use a custom version of mkdocs but external contributors don't have access to that repository. I don't know if there's a better way to support this?

**Comment by charliermarsh** [comment]
I'm not sure what you're running exactly but we need to mark Ruff as unmanaged. We don't want to be building Ruff to run scripts.

**Comment by MichaReiser** [comment]
I try to run `uv run --verbose scripts/generate_mkdocs.py`. Overall, I try to get the workflow documented in the contributing.md to work.

**Comment by charliermarsh** [comment]
It's because the workspace members aren't declared as dependencies of the project.

**Comment by charliermarsh** [comment]
`uv run --package ruff-docs --verbose scripts/generate_mkdocs.py` does the right thing, but fails because `./docs` isn't a valid Python package (it needs `src/ruff_docs/__init__.py`).

**Comment by charliermarsh** [comment]
If you add this to `docs/pyproject.toml`, then `uv run --package ruff-docs -- scripts/generate_mkdocs.py ` works:

```toml
[tool.hatch.build.targets.wheel]
packages = ["src/ruff_docs"]
```

I think we should re-add that to `uv init`, it got dropped during code review.

File: scripts/generate_mkdocs.py [hunk 3]
```diff
@@ -161,7 +161,7 @@ def main() -> None:
     # Format rules docs
     add_no_escape_text_plugin()
     for rule_doc in Path("docs/rules").glob("*.md"):
-        mdformat.file(rule_doc, extensions=["mkdocs", "admonition", "no-escape-text"])
+        mdformat.file(rule_doc, extensions=["mkdocs", "admon", "no-escape-text"])
```
Comment (charliermarsh) [inline scripts/generate_mkdocs.py:164]: This was a breaking change in `mdformat-admon`, in the 2.0.3 release: https://github.com/KyleKing/mdformat-mkdocs/commit/6a329979e27da47a215607d4e30f61d22e683b89

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -294,28 +294,30 @@ Finally, regenerate the documentation and generated code with `cargo dev generat
 
 To preview any changes to the documentation locally:
 
-1. Install the [Rust toolchain](https://www.rust-lang.org/tools/install).
+1. Install the [Rust toolchain](https://www.rust-lang.org/tools/install).]
```
Comment (charliermarsh) [inline CONTRIBUTING.md:297]: This seems off... Isn't that mismatched now?

**Comment by MichaReiser** [comment]
Okay, I don't understand why I would have to declare `ruff_docs` as a package just so that the generate script can resolve `mdformat` :confused:

**Comment by charliermarsh** [comment]
Can you expand on your question?

**Comment by MichaReiser** [comment]
It's more of an observation. I understand that the `packages` setting makes the listed package available to other packages. But the `generate_mkdocs` script doesn't depend on the `ruff_docs` package. It depends on some of its transitive dependencies. 

That's why I think it odd that I have to add `ruff_docs` to `packages` just to make the script work. I could understand if I have to move the script into the `ruff_docs` directory to make it work (so that uv knows that it should run the script with additional third party dependencies).

**Comment by charliermarsh** [comment]
@MichaReiser - I think part of the problem is that `ruff_docs` isn't really a package... An alternative is to make these all dev dependencies in the workspace root. Do you find that more intuitive?

```diff
diff --git a/pyproject.toml b/pyproject.toml
index eeba1c160..1ab117188 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -8,7 +8,7 @@ version = "0.5.6"
 description = "An extremely fast Python linter and code formatter, written in Rust."
 authors = [{ name = "Astral Software Inc.", email = "hey@astral.sh" }]
 readme = "README.md"
-requires-python = ">=3.7"
+requires-python = ">=3.8"
 license = { file = "LICENSE" }
 keywords = [
   "automation",
@@ -44,7 +44,18 @@ Documentation = "https://docs.astral.sh/ruff/"
 Changelog = "https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md"

 [tool.uv.workspace]
-members = ["docs", "python/ruff-ecosystem"]
+members = ["python/ruff-ecosystem"]
+
+[tool.uv]
+dev-dependencies = [
+  "PyYAML>=6.0.1",
+  "mkdocs>=1.5.0",
+  "mdformat>=0.7.17",
+  "mdformat-mkdocs>=2.0.4",
+  "mdformat-admon>=2.0.3",
+  "mkdocs-material>=9.1.18",
+  "mkdocs-redirects>=1.2.1",
+]

 [tool.maturin]
 bindings = "bin"
```

Then `uv run -- scripts/generate_mkdocs.py` works.

**Comment by T-256** [comment]
Can I work on this?

**Comment by MichaReiser** [comment]
@charliermarsh I see. But let's pretend that `ruff_docs` is a package. What's the recommended way to run `generate_mkdocs.py` without having to redeclare the dependencies?

@T-256 sure! Feel free to open a new PR. I then close this one

**Comment by charliermarsh** [comment]
@MichaReiser -- Genuinely trying to follow but I'm a little confused. For (2), you can add a `pyproject.toml` to `ruff_docs`, declare the dependencies in that `pyproject.toml`, and then run `uv run --package ruff-docs ./scripts/generate_mkdocs.py`. Isn't that the current state of this PR?

**Comment by MichaReiser** [comment]
>  Isn't that the current state of this PR?

Not what's described in the `CONTRIBUTING.md`. So no. I wasn't aware that this is a possibility. 

I think where the behavior of `uv run` differs from what I expected is that I expected `uv run ruff_docs/generate_mkdocs` to have the same behavior as `uv run -p ruff_docs ruff_docs/generate_mkdocs`

**Comment by charliermarsh** [comment]
Oh I see, that running a script in the `ruff_docs` directory would behave equivalently to running a script _from_ the `ruff_docs` directory.

**Pull request closed without merging**